# E07: Sample Size for a Desired Margin of Error

This notebook calculates the required sample size ($n$) to estimate a population mean with a specified margin of error and confidence level. This is a crucial step in the design of experiments.

## Theoretical Background

The calculation is based on rearranging the formula for the margin of error (E) of a confidence interval for a mean:
$$ E = z_{\alpha/2} \cdot \frac{\sigma}{\sqrt{n}} $$

Solving for $n$, we get the sample size formula:
$$ n = \left( \frac{z_{\alpha/2} \cdot \sigma}{E} \right)^2 $$

Where:
- **$E$**: The desired margin of error.
- **$\sigma$**: A planning value for the population standard deviation. Often estimated from a previous study.
- **$z_{\alpha/2}$**: The critical value from the standard normal distribution corresponding to the desired confidence level.

The final result for $n$ must always be rounded **up** to the next integer to ensure the margin of error is not exceeded.

## Step 1: Define Given Parameters

We'll use the scenario from E6, but with a new desired confidence level and margin of error as specified in E7.

In [1]:
# --- Problem Parameters ---
margin_of_error_E = 5   # Desired margin of error (in ms)
s_planning = 32         # Planning value for sigma, from E6 (in ms)
confidence_level = 0.99   # Desired confidence level

## Step 2: Determine the Critical Value (z)

For a 99% confidence interval, the significance level is $\alpha = 0.01$. The critical value $z_{\alpha/2}$ corresponds to the z-score that leaves an area of $0.005$ in the upper tail.

In [2]:
import scipy.stats as stats

alpha = 1 - confidence_level
z_critical = stats.norm.ppf(1 - alpha / 2)

print(f"Confidence Level: {confidence_level:.0%}")
print(f"Alpha (α): {alpha}")
print(f"Critical z-value (z_α/2): {z_critical:.4f}")

Confidence Level: 99%
Alpha (α): 0.010000000000000009
Critical z-value (z_α/2): 2.5758


## Step 3: Calculate the Required Sample Size (n)

Now we plug all the values into the sample size formula. The result will be a floating-point number, which we must then round up.

In [3]:
import numpy as np

# Calculate n using the formula
n_float = (z_critical * s_planning / margin_of_error_E)**2

# Round up to the next whole number, as we can't have a fraction of an observation
n_required = np.ceil(n_float)

print("--- Calculation Results ---")
print(f"Calculated n (float): {n_float:.4f}")
print(f"Required Sample Size (n, rounded up): {int(n_required)}")

--- Calculation Results ---
Calculated n (float): 271.7654
Required Sample Size (n, rounded up): 272


## Step 4: Final Conclusion

**Required Sample Size: 272**

**Conclusion:**

To achieve a 99% confidence interval for the true mean round-trip time with a margin of error of no more than $\pm5$ ms, we need to collect observations from **272 packets**.

Collecting 271 samples would result in a margin of error slightly larger than the desired 5 ms, hence the necessity of rounding up to the next integer.